File: tui_gateway/gateway_capabilities.py
"""Pure capability contract for authenticated mobile gateway clients.

This module deliberately describes protocol availability, not runtime readiness.
It must stay independent of configuration, profiles, credentials, plugins, and
host discovery so the response is deterministic and safe to expose after each
authenticated gateway connection.
"""

from collections.abc import Iterable


GATEWAY_CONTRACT_VERSION = 1
GATEWAY_MIN_COMPATIBLE = 1


# ``voice.record`` and ``voice.tts`` intentionally stay out of this mobile
# manifest: both operate audio hardware on the gateway host. Phone audio needs
# a separate transport contract before a mobile client may advertise voice.
FEATURE_METHODS: dict[str, frozenset[str]] = {
    "baseline_chat": frozenset({
        "session.create",
        "session.list",
        "session.resume",
        "prompt.submit",
    }),
    "background_work": frozenset({
        "session.active_list",
        "prompt.background",
        "session.steer",
    }),
    "delegation": frozenset({"delegation.status", "spawn_tree.list"}),
    # Repository discovery plus branch/undo is a session baseline, not the
    # structured diff/test/push/PR workflow implied by a broad "code" feature.
    "code_session_baseline": frozenset({
        "projects.discover_repos",
        "session.branch",
        "session.undo",
    }),
    "files": frozenset({"image.attach_bytes", "pdf.attach", "file.attach"}),
    "live_view": frozenset({"visual.status", "visual.frame"}),
    "handoff": frozenset({"handoff.request"}),
    "automation": frozenset({"cron.manage"}),
}


# Optional feature families are additive-optional (durable_work precedent in
# apps/mobile/contracts/gateway-feature-registry-v1.json): an omitted key means
# the family is unavailable, and the advertised boolean must always equal
# "required methods ⊆ advertised methods". Pets exposes display+adopt only;
# generation, management, and scale stay desktop/host surfaces.
OPTIONAL_FEATURE_METHODS: dict[str, frozenset[str]] = {
    "session_transcript": frozenset({"session.transcript"}),
    "pets": frozenset({
        "pet.info",
        "pet.info.meta",
        "pet.gallery",
        "pet.select",
        "pet.disable",
        "pet.thumb",
    }),
}


MOBILE_METHODS: tuple[str, ...] = tuple(
    sorted(
        set().union(*FEATURE_METHODS.values())
        | set().union(*OPTIONAL_FEATURE_METHODS.values())
        | {
            "approval.respond",
            "clarify.respond",
            "commands.catalog",
            "connection.context",
            "computer.screenshot",
            "process.kill",
            "process.list",
            "secret.respond",
            "session.close",
            "session.interrupt",
            "slash.exec",
            "sudo.respond",
        }
    )
)


def build_gateway_capabilities(
    registered_methods: Iterable[str], *, version: str, release_date: str
) -> dict:
    """Build a deterministic, redacted snapshot of the mobile RPC surface.

    ``registered_methods`` may be the live method registry or any iterable of
    method names. Only the explicitly reviewed ``MOBILE_METHODS`` allow-list is
    advertised; unreviewed registry entries can never leak through this
    boundary. Feature flags are derived from the methods required for that
    feature instead of being configured independently.
    """

    registered = frozenset(registered_methods)
    advertised = [name for name in MOBILE_METHODS if name in registered]

    return {
        "contract": {
            "name": "fabric.gateway",
            "version": GATEWAY_CONTRACT_VERSION,
            "min_compatible": GATEWAY_MIN_COMPATIBLE,
        },
        "server": {"version": version, "release_date": release_date},
        "execution": {
            "location": "gateway",
            "tool_execution": "gateway",
            "survives_client_disconnect": True,
            "survives_gateway_restart": False,
            "requires_gateway_host_online": True,
        },
        "features": {
            feature: required.issubset(registered)
            for feature, required in (
                FEATURE_METHODS | OPTIONAL_FEATURE_METHODS
            ).items()
        },
        "methods": advertised,
    }

File: apps/shared/src/gateway-capabilities.ts
import { GatewayRpcError, type JsonRpcGatewayClient } from "./json-rpc-gateway";

export const GATEWAY_CLIENT_CONTRACT_VERSION = 1;

// Gateway-host voice RPCs are deliberately absent: they record/play on the
// gateway machine, not the phone. Mobile voice needs a phone-audio contract.
export const GATEWAY_FEATURE_METHODS = {
  automation: ["cron.manage"],
  background_work: [
    "session.active_list",
    "prompt.background",
    "session.steer",
  ],
  baseline_chat: [
    "session.create",
    "session.list",
    "session.resume",
    "prompt.submit",
  ],
  code_session_baseline: [
    "projects.discover_repos",
    "session.branch",
    "session.undo",
  ],
  delegation: ["delegation.status", "spawn_tree.list"],
  files: ["image.attach_bytes", "pdf.attach", "file.attach"],
  handoff: ["handoff.request"],
  live_view: ["visual.status", "visual.frame"],
} as const;

/**
 * Additive feature gates introduced after the original version-1 fixture.
 * Their absence means "not advertised" so an older gateway remains a valid
 * version-1 peer; when present, the same method/feature invariant applies.
 */
export const OPTIONAL_GATEWAY_FEATURE_METHODS = {
  artifact_fetch: ["artifact.list", "artifact.fetch"],
  connected_nodes: ["node.list", "node.revoke"],
  device_node: ["node.enroll"],
  durable_work: [
    "job.create",
    "job.sync",
    "job.get",
    "job.list",
    "job.events",
    "job.cancel",
    "attention.get",
    "attention.list",
    "attention.respond",
  ],
  node_invoke: ["node.announce", "node.result", "node.reject"],
  push: ["push.register_device", "push.deregister_device"],
  session_admin: ["session.rename", "session.archive"],
  trust_center: [
    "trust.audit.list",
    "grant.list",
    "grant.create",
    "grant.revoke",
  ],
  workspace_read: ["fs.list", "fs.read"],
} as const;

/**
 * Optional features advertised as a bare boolean with no dedicated methods
 * (scoped_grants extends approval.respond params), so no method/feature
 * consistency check applies. Absence still means "not advertised" → false.
 */
export const OPTIONAL_GATEWAY_FEATURE_FLAGS = ["scoped_grants"] as const;

export type GatewayFeatureName =
  | keyof typeof GATEWAY_FEATURE_METHODS
  | keyof typeof OPTIONAL_GATEWAY_FEATURE_METHODS
  | (typeof OPTIONAL_GATEWAY_FEATURE_FLAGS)[number];

export const LEGACY_MOBILE_METHODS: ReadonlySet<string> = new Set([
  "approval.respond",
  "clarify.respond",
  "commands.catalog",
  "computer.screenshot",
  "process.kill",
  "process.list",
  "prompt.background",
  "prompt.submit",
  "secret.respond",
  "session.active_list",
  "session.close",
  "session.create",
  "session.interrupt",
  "session.list",
  "session.resume",
  "session.steer",
  "slash.exec",
  "sudo.respond",
]);

export interface GatewayCapabilities {
  contract: {
    min_compatible: number;
    name: "fabric.gateway";
    version: number;
  };
  execution: {
    location: "gateway";
    requires_gateway_host_online: true;
    survives_client_disconnect: true;
    survives_gateway_restart: false;
    tool_execution: "gateway";
  };
  features: Record<GatewayFeatureName, boolean>;
  methods: string[];
  server: {
    release_date: string;
    version: string;
  };
}

export type GatewayCompatibility =
  | { kind: "verified"; capabilities: GatewayCapabilities }
  | { kind: "legacy" }
  | { kind: "incompatible"; minimum: number }
  | { kind: "invalid"; message: string };

export type GatewayCapabilityClient = Pick<JsonRpcGatewayClient, "request">;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isPositiveInteger(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value) && value >= 1;
}

function invalid(message: string): GatewayCompatibility {
  return { kind: "invalid", message };
}

/**
 * Parse the process-scoped mobile gateway contract.
 *
 * Unknown fields and method names are deliberately accepted. The execution
 * truth and known feature relationships are version-1 safety invariants, so a
 * contradictory response is invalid rather than silently treated as legacy.
 */
export function parseGatewayCapabilities(raw: unknown): GatewayCompatibility {
  if (!isRecord(raw)) {
    return invalid("Gateway capabilities must be an object.");
  }

  const contract = raw.contract;
  if (!isRecord(contract)) {
    return invalid("Gateway capabilities are missing a contract object.");
  }
  if (contract.name !== "fabric.gateway") {
    return invalid("Gateway capability contract name must be fabric.gateway.");
  }
  if (!isPositiveInteger(contract.version)) {
    return invalid(
      "Gateway capability contract version must be a positive integer.",
    );
  }
  if (!isPositiveInteger(contract.min_compatible)) {
    return invalid(
      "Gateway minimum compatible version must be a positive integer.",
    );
  }
  if (contract.min_compatible > contract.version) {
    return invalid(
      "Gateway minimum compatible version cannot exceed its contract version.",
    );
  }

  const server = raw.server;
  if (
    !isRecord(server) ||
    typeof server.version !== "string" ||
    !server.version.trim() ||
    typeof server.release_date !== "string" ||
    !server.release_date.trim()
  ) {
    return invalid("Gateway capabilities contain invalid server metadata.");
  }

  const execution = raw.execution;
  if (!isRecord(execution)) {
    return invalid("Gateway capabilities are missing execution semantics.");
  }
  if (
    execution.location !== "gateway" ||
    execution.tool_execution !== "gateway" ||
    execution.survives_client_disconnect !== true ||
    execution.survives_gateway_restart !== false ||
    execution.requires_gateway_host_online !== true
  ) {
    return invalid(
      "Gateway capabilities contradict the version-1 execution contract.",
    );
  }

  if (!Array.isArray(raw.methods)) {
    return invalid("Gateway capability methods must be an array.");
  }
  const methods: string[] = [];
  const methodSet = new Set<string>();
  for (const method of raw.methods) {
    if (typeof method !== "string" || !method.trim()) {
      return invalid("Gateway capability methods must be non-empty strings.");
    }
    if (methodSet.has(method)) {
      return invalid(`Gateway capability method is duplicated: ${method}.`);
    }
    methods.push(method);
    methodSet.add(method);
  }

  const rawFeatures = raw.features;
  if (!isRecord(rawFeatures)) {
    return invalid("Gateway capabilities are missing feature availability.");
  }
  const features = {} as Record<GatewayFeatureName, boolean>;
  for (const feature of Object.keys(
    GATEWAY_FEATURE_METHODS,
  ) as (keyof typeof GATEWAY_FEATURE_METHODS)[]) {
    const value = rawFeatures[feature];
    if (typeof value !== "boolean") {
      return invalid(`Gateway feature ${feature} must be a boolean.`);
    }
    const expected = GATEWAY_FEATURE_METHODS[feature].every((method) =>
      methodSet.has(method),
    );
    if (value !== expected) {
      return invalid(
        `Gateway feature ${feature} contradicts its advertised methods.`,
      );
    }
    features[feature] = value;
  }
  for (const feature of Object.keys(
    OPTIONAL_GATEWAY_FEATURE_METHODS,
  ) as (keyof typeof OPTIONAL_GATEWAY_FEATURE_METHODS)[]) {
    const value = rawFeatures[feature];
    const required = OPTIONAL_GATEWAY_FEATURE_METHODS[feature];
    const expected = required.every((method) => methodSet.has(method));
    if (value === undefined) {
      features[feature] = false;
      continue;
    }
    if (typeof value !== "boolean") {
      return invalid(`Gateway feature ${feature} must be a boolean.`);
    }
    if (value !== expected) {
      return invalid(
        `Gateway feature ${feature} contradicts its advertised methods.`,
      );
    }
    features[feature] = value;
  }
  for (const feature of OPTIONAL_GATEWAY_FEATURE_FLAGS) {
    const value = rawFeatures[feature];
    if (value === undefined) {
      features[feature] = false;
      continue;
    }
    if (typeof value !== "boolean") {
      return invalid(`Gateway feature ${feature} must be a boolean.`);
    }
    features[feature] = value;
  }

  if (contract.min_compatible > GATEWAY_CLIENT_CONTRACT_VERSION) {
    return { kind: "incompatible", minimum: contract.min_compatible };
  }

  return {
    kind: "verified",
    capabilities: {
      contract: {
        min_compatible: contract.min_compatible,
        name: "fabric.gateway",
        version: contract.version,
      },
      execution: {
        location: "gateway",
        requires_gateway_host_online: true,
        survives_client_disconnect: true,
        survives_gateway_restart: false,
        tool_execution: "gateway",
      },
      features,
      methods,
      server: {
        release_date: server.release_date,
        version: server.version,
      },
    },
  };
}

export async function negotiateGatewayCapabilities(
  client: GatewayCapabilityClient,
): Promise<GatewayCompatibility> {
  try {
    const raw = await client.request<unknown>("gateway.capabilities");
    return parseGatewayCapabilities(raw);
  } catch (error) {
    if (
      error instanceof GatewayRpcError &&
      error.kind === "rpc" &&
      error.code === -32601
    ) {
      return { kind: "legacy" };
    }
    throw error;
  }
}

/**
 * Whether a feature family is usable on this gateway. Mirrors the durable_work
 * precedent: an optional family only exists when a verified contract advertises
 * it true — a legacy gateway predates every optional family, and an
 * incompatible or invalid contract fails closed, even when the raw method
 * names would appear to overlap.
 */
export function supportsGatewayFeature(
  state: GatewayCompatibility,
  feature: GatewayFeatureName,
): boolean {
  return (
    state.kind === "verified" && state.capabilities.features[feature] === true
  );
}

export function supportsGatewayMethod(
  state: GatewayCompatibility,
  method: string,
): boolean {
  if (state.kind === "verified") {
    return state.capabilities.methods.includes(method);
  }
  return state.kind === "legacy" && LEGACY_MOBILE_METHODS.has(method);
}
